python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+    # *************************************************************************
+    @classmethod
+    def _can_be_null(cls, cspec):
+        """Return true if this column may have a null value.
+
+        Parameters
+        ----------
+        cspec : 'string'
+            The tokenized line from the schema file corresponding to a column.
+
+        Returns
+        -------
+        result : 'bool'
+             Returns True if the column may contain a null value, False
+             otherwise.
+
+        Notes
+        -----
+        - This method uses the common SQL syntax that a schema line must
+          contain 'not null' for the column not be able to have a null
+          value. Also, it is assumed that comments have been removed from
+          the schema definition.
+        """
+
+        # Return True is 'not null' is not present
+        #
+        return 'NOT NULL' not in ' '.join(cspec).upper()
+
+    # *************************************************************************
+    # *                      d i s p l a y _ c s v 2 p q                      *
+    # *************************************************************************
+    @classmethod
+    def display_csv2pq(cls):
+        """Display conversion of csv file to this parquet schema.
+
+        Notes
+        -----
+        - The format used is:
+          <colno> <colname> <sqltype> -> <pqtype>
+        """
+
+        # Check if we should display the scheme
+        #
+        if Schema._colSpecs:
+            n = len(str(len(Schema.colNames)))
+            fmt1 = '{0: >' + str(n) + '} {1: <' + str(Schema._cNameMax) +\
+                   '} {2}'
+            fmt2 = fmt1 + ' -> {3}'
+            n = -1
+            for cname, sqltype, nptype in Schema._colSpecs:
+                n += 1
+                if not sqltype or sqltype == nptype.upper():
+                    print(fmt1.format(n, cname, nptype))
+                else:
+                    print(fmt2.format(n, cname, sqltype, nptype))
+            if Schema.colOrigs < len(Schema.colNames):
+                for cname in Schema.colNames[Schema.colOrigs:]:
+                    n += 1
+                    print(fmt1.format(n, cname, "bool"))
+
+    # *************************************************************************
+    # *                      d i s p l a y _ p q 2 c s v                      *
+    # *************************************************************************
+    @classmethod
+    def display_pq2csv(cls):
+        """Display conversion of parquet file to this csv schema.
+
+        Notes
+        -----
+        - It is assumed that the caller has verified that the underlying
+          parquet file corresponds to this schema.
+        - The format used is:
+          <colno> <colname> <pqtype> -> <csvtype>
+        """
+
+        # Setup to format the schmema conversion
+        #
+        if Schema.colOrigs:
+            pqcols = Schema._colSpecs[0:Schema.colOrigs]
```

Comment:
I think you don't need the 0 when slicing the list?

Reply:
Correct, the default is zero. This is a matter of non-prohibited style so I am not disposed to changing it.